src/storage/txn/commands/flashback_to_version_read_phase.rs

```diff
@@ -59,41 +77,72 @@ impl<S: Snapshot> ReadCommand<S> for FlashbackToVersionReadPhase {
             (key_locks, has_remain_locks) = key_locks_result?;
         }
         // Scan the writes.
-        let mut key_writes = Vec::with_capacity(0);
+        let mut keys = Vec::with_capacity(0);
         let mut has_remain_writes = false;
         // The batch is not full, we can still read.
         if self.next_write_key.is_some() && key_locks.len() < FLASHBACK_BATCH_SIZE {
+            let cur_key = RefCell::new(None);
             let key_writes_result = reader.scan_writes(
                 self.next_write_key.as_ref(),
                 self.end_key.as_ref(),
-                // To flashback `CF_WRITE` and `CF_DEFAULT`, we need to delete all keys whose
-                // commit_ts is greater than the specified version.
-                |key| key.decode_ts().unwrap() > self.version,
+                // To flashback the data, we need to get all the latest writes first by scanning
+                // every unique key in `CF_WRITE`.
+                |key| {
+                    // No need to flashback the key if its `commit_ts` is smaller than
+                    // `self.version`.
+                    if let Ok(commit_ts) = key.decode_ts() {
+                        if commit_ts <= self.version {
+                            return false;
+                        }
+                    }
+                    // Only scan the latest version for each key, i.e, the first version we meet
+                    // for each key.
+                    if let Ok(truncated_key) = key.clone().truncate_ts() {
+                        let mut cur_key_rc = cur_key.borrow_mut();
+                        if cur_key_rc.is_some() && truncated_key == *cur_key_rc.as_ref().unwrap() {
+                            return false;
```

Comment:
If the two versions of the same key are read in different batches, `cur_key` checking would miss it.